File: 设计模式/程序/2-AbstractFactoryPattern/AbFactoryPattern(C#)/ColorFactory.cs
/* 颜色工厂 */
using System;


namespace AbstractFactoryPattern
{
    public interface Color
    {
        void fill();
    }

    public class Red : Color
    {
        public void fill()
        {
            Console.WriteLine("Fill with Red!");
        }
    }

    public class Green : Color
    {
        public void fill()
        {
            Console.WriteLine("Fill with Green");
        }
    }

    /* 颜色的工厂类 */
    public class ColorFactory : AbstractFactory
    {   
        public Color GetColor(string colorType)
        {
            if (colorType == "RED")
                return new Red();
            else if (colorType == "GREEN")
                return new Green();
            else 
                return null;
        }
        public Shape GetShape(string shapeType)
        {
            return null;
        }
    }
}

File: 设计模式/程序/2-AbstractFactoryPattern/AbFactoryPattern(Cpp)/src/ColorFactory.cpp
#include "ColorFactory.h"
#include "ShapeFactoey.h"
#include <iostream>
#include <string>


namespace AbstratFactoryPattern
{
    void Red::fill()
    {
        std::cout << "fill with Red" << std::endl;
    }

    void Green::fill()
    {
        std::cout << "fill with Green" << std::endl;
    }
    

    /**********************************************************
     * @brief ColorFactory's GetColor
     * @param colorType color need to create
     * @return Color* object point
     **********************************************************/
    Color* ColorFactory::GetColor(const std::string& colorType)
    {
        if (colorType == "Red")
            return new Red();
        else if (colorType == "Green")
            return new Green();
        return nullptr;
    }

    /**********************************************************
     * @brief ColorFactory's GetShape
     * @param shapeType shape need to create
     * @return Color* object point
     **********************************************************/
    Shape* ColorFactory::GetShape(const std::string& shapeType)
    {
        return nullptr;
    }
}